lib/bootsnap/cli.rb

```diff
@@ -19,59 +19,62 @@ def match?(string)
 
     attr_reader :cache_dir, :argv
 
-    attr_accessor :compile_gemfile, :exclude
+    attr_accessor :compile_gemfile, :exclude, :verbose, :iseq, :yaml
 
     def initialize(argv)
       @argv = argv
       self.cache_dir = ENV.fetch('BOOTSNAP_CACHE_DIR', 'tmp/cache')
       self.compile_gemfile = false
       self.exclude = nil
+      self.verbose = false
+      self.iseq = true
+      self.yaml = true
     end
 
     def precompile_command(*sources)
       require 'bootsnap/compile_cache/iseq'
+      require 'bootsnap/compile_cache/yaml'
 
       fix_default_encoding do
         Bootsnap::CompileCache::ISeq.cache_dir = self.cache_dir
+        Bootsnap::CompileCache::YAML.init!
+        Bootsnap::CompileCache::YAML.cache_dir = self.cache_dir
 
-        if compile_gemfile
-          sources += $LOAD_PATH
-        end
+        main_sources = sources.map { |d| File.expand_path(d) }
+        precompile_ruby_files(main_sources)
+        precompile_yaml_files(main_sources)
 
-        sources.map { |d| File.expand_path(d) }.each do |path|
-          if !exclude || !exclude.match?(path)
-            list_ruby_files(path).each do |ruby_file|
-              if !exclude || !exclude.match?(ruby_file)
-                CompileCache::ISeq.fetch(ruby_file, cache_dir: cache_dir)
-              end
-            end
-          end
+        if compile_gemfile
+          # Some gems embed their tests, they're very unlikely to be loaded, so not worth precompiling.
+          gem_exclude = Regexp.union([exclude, %r{/spec/}, %r{/test/}].compact)
+          precompile_ruby_files($LOAD_PATH.map { |d| File.expand_path(d) }, exclude: gem_exclude)
+          precompile_yaml_files($LOAD_PATH.flat_map { |p| p.scan(/.*\/gems\/[^\/]+\//) }.uniq, exclude: gem_exclude)
```

Comment:
For this one the `%r{}` syntax would be useful.  I think `scan` is a bit weird, I don't think we expect multiple results from a single load path.

```suggestion
          precompile_yaml_files($LOAD_PATH.map {|p| p[%r{.*/gems/[^/]+/}] }.compact.uniq, exclude: gem_exclude)
```

Also `/gems/` is [definitely in the path](https://github.com/rubygems/rubygems/blob/b5a113c040a1820a9f0fdd9a11965e695939e2be/lib/rubygems/installer.rb#L262), but technically one of the gem paths could have a /gems/ before the actual gems directory. Maybe we could go off `Gem.path` instead?

Finally the idea of going at the gem top-level dir instead of in the load path is because sometimes gems store YAML files in dirs that are not in the load path like `config/`? Could we add a test or two for this line somehow?

Reply:
> I think scan is a bit weird, I don't think we expect multiple results from a single load path.

We indeed don't. But it synergize surprisingly well with `flat_map`, I somehow find it elegant. But maybe a `String.sub` might be simpler. I'll dig that a bit.

> Maybe we could go off Gem.path instead?

Hum, yeah I'll check that. The `/gem/` is indeed a bit too hacky.

> Finally the idea of going at the gem top-level dir instead of in the load path is because sometimes gems store YAML files in dirs that are not in the load path like config/?

Yes, exactly.

> Could we add a test or two for this line somehow?

I'll see about that.

Reply:
yeah I'm not a fan of map/compact either, I would have gone with filter_map but I think we want to keep support for 2.6. But other than that `String#[](Regexp)` is kind of a single element `String#scan(Regexp)`, so it fits well. But it's just nitpicks anyway, whatever works best for you.